### ISSUE TITLE
catch suspicious clones: let foo = foo.clone();   to appease borrowck

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#5904]
Inspired by this change: https://github.com/rust-lang/rust-clippy/pull/5897/files

It might be interesting to catch occurrences of `let foo = foo.clone()` in order to make code work where we get a reference of foo `&foo` but need to pass it to a function that consumes `foo` and `.clone()` it to appease the borrow checker:

````rust
 #[derive(Clone)]
struct A {}

fn main() {
    let a: A = A {};
    no_consume(&a);
}

fn consume(a: A) {}

fn no_consume(a: &A) {
    let a = a.clone();
    consume(a);
}
````

IMO code like this is worth looking at to see if we can avoid the clone by not passing &A by reference:

````rust
fn main() {
    let a: A = A {};
    do_consume(a);
}

fn consume(a: A) {}

fn do_consume(a: A) {
    consume(a);
}
````

There might be a lot of false positives so this would probably be a pedantic lint.
The lint should probably not fire for `let mut foo = foo.clone()` as well?

**Comment by HMPerson1** [comment]
Instead of matching variable names, would be better to check for parameters whose only use is calling `.clone()`? I think it would be more precise and it would still catch this example.

**Comment by camsteffen** [comment]
This is basically the opposite of `needless_pass_by_value`. Good idea. The problem is that the `clone` may be seen as implementation detail that doesn't justify changing the function signature. Or it may be impractical to `clone` for every usage of the function.
